chore(tutorial): Using multiple themes together

ehowey opened the pull request:
## Description

A tutorial for composing multiple themes together. This is my first bigger PR - I apologize if I miss protocol with resolving comments, feel free to give me direction if I do something wrong from a git workflow/review process!

Progress:

- Completed a first draft
- Demo repo is here: https://github.com/ehowey/gatsby-themes-tutorial, I have tried to break each step down into its own commit.

Questions:

- I put in a pretty big navigation component using shadowing at the end.  This may be unnecessary but I think it helps to illustrate the power of shadowing in greater detail. Do you think we should keep this large example or is it too complex?
- I originally had more 'voice' in the writing but tried to remove some of that so it was more technical and to the point. Perhaps I need to remove even more personality from it?
- Do we want or need pictures/images of the theme progress?

Laurie:
- Those issues have now all been resolved from my original post.  I decided to just include a more complete siteMetadata example in the tutorial, thank you for such a detailed and speedy reply! The issue with NPM was my end.  It installed with yarn first then I tried to mix in NPM, which is why I had a problem.  I just had to do a fresh install with NPM from the start. It sounds like you folks are already looking at theme queries and 'robustness' of those queries!

Obinna:
- Thanks for some of those first reviews and comments! I think I got all those covered with the updated work!

ehowey commented:
This is far from done but I am running into some issues as I work through this tutorial that are things that need to be fixed/checked first. I will track my progress in here.

Current issues:

- gatsby-theme-notes has a bug, see https://github.com/gatsbyjs/gatsby/issues/21713
- npm install was not working, yarn was working. I have to check into why this is
- Fragile GraphQL queries were causing failures that should not happen. E.g. the field "social" was non-nullable, but really should be or a default should be provided.  I will do some digging on this and see what I unearth. This will require some work on the themes themselves. In my experience with themes this is where it gets a bit tricky when composing themes.

laurieontech commented:
Thanks for the work @ehowey! I haven't reviewed the PR but wanted to add some context on your notes.
1. The bug is now fixed! Yay!
2. npm install should work, so let's dig into that a bit more
3. The README for blog theme is currently missing some information, including that social is non-nullable. We're working on some other updates to themes fragility, but I'll note the possibility of a default for those fields.
4. I think you're right to use blog-theme instead of blog-theme-core. As you mentioned, the alternative would be much longer.

Ekwuno commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,247 @@
+- learned about component shadowing
+- learned about shadowing a Theme-UI theme file
+
+The themes we will be composing in this tutorial are [gatsby-theme-blog](https://www.gatsbyjs.org/packages/gatsby-theme-blog/), [gatsby-theme-notes](https://www.gatsbyjs.org/packages/gatsby-theme-notes/) and [gatsby-mdx-embed](https://gatsby-mdx-embed.netlify.com/). Follow along as we go from 'hello-world' to a fully functioning final site!
```
```suggestion
The themes we will be composing in this tutorial are [gatsby-theme-blog](/packages/gatsby-theme-blog/), [gatsby-theme-notes](/packages/gatsby-theme-notes/) and [gatsby-mdx-embed](/packages/@pauliescanlon/gatsby-mdx-embed/). Follow along as we go from 'hello-world' to a fully functioning final site!
```
Inside docs we can use relative links (/packages/gatsby-theme-blog/ instead of https://www.gatsbyjs.org/packages/gatsby-theme-notes/), which utilize the Gatsby Link component

ehowey replied:
Done, I think I caught them all!

Ekwuno commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,247 @@
 ---
 
-_This tutorial is coming soon. In the meantime, check out the [introduction to using multiple themes in the docs](/docs/themes/using-multiple-gatsby-themes)._
+## What’s contained in this tutorial?
```
```suggestion
## What’s contained in this tutorial?
```
## What this tutorial covers:

Ekwuno commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,247 @@
+- learned about component shadowing
+- learned about shadowing a Theme-UI theme file
+
+The themes we will be composing in this tutorial are [gatsby-theme-blog](https://www.gatsbyjs.org/packages/gatsby-theme-blog/), [gatsby-theme-notes](https://www.gatsbyjs.org/packages/gatsby-theme-notes/) and [gatsby-mdx-embed](https://gatsby-mdx-embed.netlify.com/). Follow along as we go from 'hello-world' to a fully functioning final site!
```
```suggestion
The themes we will be composing in this tutorial are [gatsby-theme-blog](https://www.gatsbyjs.org/packages/gatsby-theme-blog/), [gatsby-theme-notes](https://www.gatsbyjs.org/packages/gatsby-theme-notes/) and [gatsby-mdx-embed](https://gatsby-mdx-embed.netlify.com/). 
```

I think the last line can be done without. I might be wrong, what are your thoughts?

ehowey replied:
Cleaned this section up!

Ekwuno commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,481 @@
+
+## Step 3: Update gatsby-config.js
+
+The themes need to be added to the plugins array and we need some basic site metadata too.
```
```suggestion
The themes need to be added to the plugin array and the site metadata that comes with a starter should be renamed for personalization.
```
In the doc try to avoid using the first person “I”, “we”, “let’s”, and “us” per the [style guide](https://www.gatsbyjs.org/contributing/gatsby-style-guide/)

Ekwuno commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,481 @@
+
+## Step 5: Add some content
+
+Not much to look at yet, just a plain `hello-world`, but it works! Behind the scenes our two themes [created content folders](/docs/themes/conventions/#initializing-required-directories) for us in the root directory of the site. We need to add some content to these folders.
```
```suggestion
Behind the scenes, the two themes created content folders in the root directory of the site. In this step, you will add some content to these folders.
```
Try to avoid extra links when writing beginner-friendly tutorials as it might be a lot of distractions.

Ekwuno commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,481 @@
+
+## Step 11: Add a navigation menu
+
+We can use component shadowing one more time to add a basic navigation menu. You can read more about [creating dynamic navigation menus](/docs/creating-dynamic-navigation/) in the docs.
```
```suggestion
Implementing component shadowing one more time to add a basic navigation menu can be an option. You can read more about [creating dynamic navigation menus](/docs/creating-dynamic-navigation/) in the docs.
```

Ekwuno commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,481 @@
+
+## Step 9: Shadow the Theme-UI theme file
+
+`gatsby-theme-blog` and `gatsby-theme-notes` both use [Theme-UI](https://www.gatsbyjs.org/docs/theme-ui/) design tokens to manage their styling; colors, font sizes, spacing, etc. We can use component shadowing to gain control over these design tokens in our final site. Don't forget to stop and restart your development server the again. Your file structure should look like this:
```
```suggestion
`gatsby-theme-blog` and `gatsby-theme-notes` both use [Theme-UI](/docs/theme-ui/) design tokens to manage their styling; colors, font sizes, spacing, etc. We can use component shadowing to gain control over these design tokens in the final site. Don't forget to stop and restart your development server the again. Your file structure should look like this:
```

Ekwuno commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,481 @@
+
+## Step 8: Shadow the 'bio-content.js' component
+
+Your name probably isn't Jane Doe. Let's fix that with a custom `bio-content.js` component using [theme shadowing](/docs/themes/shadowing/). Don't forget to stop and restart your development server when adding a shadowed component for the first time. Your file structure should look like this:
```
```suggestion
Your name probably isn't Jane Doe. It is possible to update this information with a custom `bio-content.js` component using [theme shadowing](/docs/themes/shadowing/). 

> Don't forget to stop and restart your development server when adding a shadowed component for the first time. 

Your file structure should look like this:
```

ehowey commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,481 @@
+
+- [Gatsby fundamentals](/tutorial/#gatsby-fundamentals)
+- [What are Gatsby themes?](/docs/themes/what-are-gatsby-themes/)
+
```
I added an "Overview" section at the start, let me know if that is what you meant since it is a longer tutorial.

Ekwuno replied:
Looks great! Thanks. It would be great to be able to link the sections to this overview. What do you think?

muescha replied:
maybe use the TOC element here?

ehowey commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,481 @@
+- Component shadowing for Theme-UI theme file
+
+## Prerequisites
+
```
Did you mean to add a link to the reference guide in the prerequisites section as well? I added a link at the end under other resources.  My sense is that second link to "What are Gatsby themes" would lead folks to the reference guide easily if they wanted.

Ekwuno replied:
Oh if you have a link to that in the other resources section, then there is no need for it here my bad. I would ask that the other changes be implemented, I think it reads better.  what are your thoughts?

ehowey commented:
@Ekwuno I think I got those changes in you suggested.  I collapsed a couple of the steps in the tutorial to try and group things together, e.g. "Adding Content" is now all one step instead of split into two steps. I caught another couple of language/grammar items and tightened up a few other areas, trying to be more concise. This is great keep the feedback coming!

muescha commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,508 @@
+        url: `https://twitter.com/gatsbyjs`,
+      },
+      {
+        name: `Github`,
```
```suggestion
        name: `GitHub`,
```

muescha replied:
```suggestion
        name: `GitHub`,
```

muescha commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,508 @@
+    ],
+    social: [
+      {
+        name: `twitter`,
```
```suggestion
        name: `Twitter`,
```

muescha commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,508 @@
+        url: `https://twitter.com/gatsbyjs`,
+      },
+      {
+        name: `github`,
```
```suggestion
        name: `GitHub`,
```

Ekwuno commented:
Thanks for implementing the changes Eric, I will test this tutorial today.

muescha commented on docs/tutorial/using-multiple-themes-together.md:
````diff
@@ -2,4 +2,508 @@
+
+Your file structure should look like this:
+
+```
````
````suggestion
```text
````

muescha commented on docs/tutorial/using-multiple-themes-together.md:
````diff
@@ -2,4 +2,508 @@
+
+Your file structure should look like this:
+
+```
````
````suggestion
```text
````

muescha commented on docs/tutorial/using-multiple-themes-together.md:
````diff
@@ -2,4 +2,508 @@
+
+Shadow `header.js` from `gatsby-theme-blog`:
+
+```
````
````suggestion
```test
````

muescha commented on docs/tutorial/using-multiple-themes-together.md:
````diff
@@ -2,4 +2,508 @@
+
+Test it out by adding a Youtube video to one of your blog posts:
+
+```md:title=content/posts/hello-posts.mdx
````
````suggestion
```mdx:title=content/posts/hello-posts.mdx
````
filenames are `mdx` so the language should all also `mdx`?

ehowey commented:
@Ekwuno Just wondering if you had a chance to test this?
@muescha Thank you die another set of eyes on this!

Ekwuno commented:
Hi @ehowey, I apologise for the delay. I should get this tested tomorrow.  Thanks for your patience with me.

laurieontech commented:
This is really great! Since some of these code examples are quite large, I'm wondering if it would make sense to add an example project into the examples/ directory so we can link to the appropriate sections and have a full working project?

ehowey commented:
@Ekwuno Thanks for all of your help and support! I am glad it all worked!

@laurieontech That does sound like a good idea to have an example.  I tracked my tutorial pretty closely in this repo:
https://github.com/ehowey/gatsby-themes-tutorial

Could we just take the code from that and use it as an example?

ehowey commented:
> Tested! Looks good. Thanks for your patience with me

One question, I noticed you were specifying localhost:8001 in the docs, was there a reason for this?

muescha commented:
i would also change it to the default port `8000`

Ekwuno commented:
> This is really great! Since some of these code examples are quite large, I'm wondering if it would make sense to add an example project into the examples/ directory so we can link to the appropriate sections and have a full working project?

I agree that'd be awesome! I have the tutorial set up locally and I will update the examples directory and link it on Monday.

Ekwuno commented:
> > Tested! Looks good. Thanks for your patience with me
> 
> One question, I noticed you were specifying localhost:8001 in the docs, was there a reason for this?

Oh no! I had another development server running. I didn't realize that affected the tutorial. As @muescha suggested port `8000` works!

Ekwuno commented:
> @Ekwuno Thanks for all of your help and support! I am glad it all worked!
> 
> @laurieontech That does sound like a good idea to have an example. I tracked my tutorial pretty closely in this repo:
> [ehowey/gatsby-themes-tutorial](https://github.com/ehowey/gatsby-themes-tutorial)
> 
> Could we just take the code from that and use it as an example?

Oh yes, you can do that. Navigate to the `/gatsby/examples` directory and create a folder. maybe call it "using-multiple-theme" then add the code from  [ehowey/gatsby-themes-tutorial](https://github.com/ehowey/gatsby-themes-tutorial) and link to the appropriate sections. Let me know if you have any questions @ehowey

ehowey commented:
Sorry @Ekwuno I will try to get to this tonight - been caught up in all of this COVID-19 stuff. My apologies.

Ekwuno commented:
> Sorry @Ekwuno I will try to get to this tonight - been caught up in all of this COVID-19 stuff. My apologies.

Oh no! I totally get it. Thanks for doing this. I already got started on the example site `README.md`. I will just push these changes. I appreciate your work on this.

laurieontech commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,530 @@
+- That you have an understanding of [Gatsby fundamentals](/tutorial/#gatsby-fundamentals)
+- An existing knowledge of [Gatsby themes](/docs/themes/what-are-gatsby-themes/)
+
+## Step 1: Create a new site using the hello world starter and change directory to "multiple-themes"
```
```suggestion
## Create a new site 

Use the hello world starter to create a new site and navigate into the "multiple-themes" directory.
```

laurieontech commented on docs/tutorial/using-multiple-themes-together.md:
````diff
@@ -2,4 +2,530 @@
+cd multiple-themes
+```
+
+## Step 2: Install and compose the themes
````
```suggestion
## Install and compose the themes
```

laurieontech commented on docs/tutorial/using-multiple-themes-together.md:
````diff
@@ -2,4 +2,530 @@
+}
+```
+
+## Step 3: Run the site
````
```suggestion
## Run the site
```

laurieontech commented on docs/tutorial/using-multiple-themes-together.md:
````diff
@@ -2,4 +2,530 @@
+cd multiple-themes
+gatsby develop
+```
+
````
Wondering if this should be its own step since it's sort of an in between step? Perhaps we could put it as part of the previous section and note that you'll want to run your site periodically to check on your changes.

laurieontech commented on docs/tutorial/using-multiple-themes-together.md:
````diff
@@ -2,4 +2,530 @@
+gatsby develop
+```
+
+## Step 4: Add content
````
```suggestion
## Add content
```

laurieontech commented on docs/tutorial/using-multiple-themes-together.md:
````diff
@@ -2,4 +2,530 @@
+    },
+```
+
+## Step 6: Shadow the 'bio-content.js' component
````
```suggestion
## Shadow the 'bio-content.js' component
```

laurieontech commented on docs/tutorial/using-multiple-themes-together.md:
````diff
@@ -2,4 +2,530 @@
+)
+```
+
+## Step 7: Shadow the Theme-UI theme file
````
```suggestion
## Shadow Theme-UI
```

laurieontech commented on docs/tutorial/using-multiple-themes-together.md:
````diff
@@ -2,4 +2,530 @@
+<YouTube youTubeId="6Z4p-qjnKCQ" />
+```
+
+## Step 9: Use component shadowing to add a navigation menu
````
```suggestion
## Add a navigation menu
```

ehowey commented on docs/tutorial/using-multiple-themes-together.md:
````diff
@@ -2,4 +2,530 @@
+- Change the theme options for the blog theme in `gatsby-config.js`.
+- Restart your development server with `gatsby develop`.
+
+```javascript:title=gatsby-config.js
````
I hummed and hawed about this when writing the first draft. This keeps the "option" visible and I prefer it in my own coding style - going back to a single line install obfuscates the power of theme options.  That is my 2 cents, but happy to revert to the single line install.

laurieontech replied:
That seems reasonable for clarity. Perhaps we can add a note that mentions it could revert to a single line?

ehowey commented on docs/tutorial/using-multiple-themes-together.md:
````diff
@@ -2,4 +2,530 @@
+      },
+    },
+```
+
````
It is mentioned just above in the bullet points, is that what you meant?

laurieontech replied:
Ah oops, missed that while going through. We could move that after the code sample.

ehowey replied:
Done - see incoming changes.

ehowey commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,530 @@
+
+## Step 5: Put the blog posts on the homepage
+
+- Stop your development server, `CTRL + C`.
```
Agree - I went back and forth on this as well.  I just didn't want someone to be "why isn't this working!!" not realizing they need to stop and start again.

ehowey commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,530 @@
+
+Put an avatar image into `content/assets/` directory
+
+Add at least one post and one note:
```
Done, see incoming changes

ehowey commented:
@laurieontech @Ekwuno - I got part way through changes, but it is movie and relax night here at home and I need to focus on that.  Obinna - I saw you started building an example repo.  I just copy and pasted what I had in, I haven't looked over it or tested it.  Feel free to hack away at what I put in there.

Thank-you for the kind words and help with this!  Have a good night.  I will get back to it tomorrow I hope.

laurieontech commented:
@ehowey We really appreciate all your work on this and absolutely take whatever time you need. Family comes first.

laurieontech commented:
Also, looks like there are two files that got committed that maybe shouldn't have been? www/src/data/locales/en/messages.po and another locales change.

ehowey commented:
> Also, looks like there are two files that got committed that maybe shouldn't have been? www/src/data/locales/en/messages.po and another locales change.

Not sure how I changed those but I think I got the changes out!  Sorry about that.

laurieontech commented:
Git is weird sometimes. Looks like they're gone now :)

muescha commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,527 @@
+- [Setting up yarn workspaces for Gatsby theme development](/blog/2019-05-22-setting-up-yarn-workspaces-for-theme-development/#reach-skip-nav)
+- [What is component shadowing?](/blog/2019-04-29-component-shadowing/)
+- [Customizing styles in Gatsby themes with Theme-UI](/blog/2019-07-03-customizing-styles-in-gatsby-themes-with-theme-ui/)
+- [Composing and styling Gatsby themes (with Brent Jackson) — Learn With Jason](https://www.youtube.com/watch?v=6Z4p-qjnKCQ)
```
```suggestion
- [Composing and styling Gatsby Themes (with Brent Jackson) — Learn With Jason](https://www.youtube.com/watch?v=6Z4p-qjnKCQ)
```

muescha commented on examples/using-multiple-themes/README.md:
```diff
@@ -0,0 +1,60 @@
+<p align="center">
+  <a href="https://www.gatsbyjs.org">
+    <img alt="Gatsby logomark" src="https://www.gatsbyjs.org/monogram.svg" width="60" />
```
should the logomark not better in the example repo?

ehowey replied:
I'm going to leave this for Laurie or Obinna.

ehowey commented on docs/tutorial/using-multiple-themes-together.md:
````diff
@@ -2,4 +2,530 @@
+`src/gatsby-plugin-theme-ui/index.js`
+and your component could look like this:
+
+```javascript:title=index.js
````
Hahaha - yes those colors were totally random picks of basically the first css colors that came to my mind I will put in some more reasonable ones in the next commit.  I think a few screenshots is a great idea - could someone from the learning team take that on? I noticed in the "Using A Theme" tutorial there are some nice screenshots with a cool Gatsby border on them.

ehowey commented on docs/tutorial/using-multiple-themes-together.md:
````diff
@@ -2,4 +2,530 @@
+```
+
+Update the `gatsby-config.js` file and add `gatsby-mdx-embed` as a plugin:
+
````
Agree - see incoming changes.

ehowey commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,530 @@
+        url: `https://github.com/gatsbyjs`,
+      },
+    ],
+    // highlight-end
```
Yep - see incoming changes.

ehowey commented:
@laurieontech I think I got all your changes incorporated in. Picked some better colors that passed A11y contrast checks. I also added a "What's Next?" section to Using A Theme as well.  I am not sure if I incorporated the example repo in the right way.  Feel free to just go in there and make changes to how I did it. I shortened and abbreviated a couple of code examples as well. Oh and I added a small "Wrapping Up" section similar to Using A Theme.

I think that is it...pretty happy with where it is at now. Let me know if there are additional changes needed.

laurieontech commented:
Thanks @ehowey! I'll take a look.

laurieontech commented:
I pulled this down locally and made a few final adjustments but I think this is mostly good to go. One more team member is taking a look and then we should be able to merge this!

ehowey commented:
One thing I ran into over the weekend that we may want to include. Theme-UI has its own merge algorithm now. See: https://theme-ui.com/guides/merging-themes

I am not at a computer but we could use this instead of deep merge. Would want to just quickly test it. But maybe worth including this in the example of shadowing Theme-UI.

laurieontech commented:
Sure @ehowey, that sounds like a good plan. Do you want to do that or would you like me to?

ehowey commented:
That would be great if you did it! I am away from a computer for the day.

LekoArts commented on docs/tutorial/using-multiple-themes-together.md:
```diff
@@ -2,4 +2,410 @@
+
+4. Import the navigation menu and add it to the header:
+
+> 💡 This code snippet is edited for length the [full component can be viewed on GitHub.](/examples/using-multiple-themes/src/gatsby-theme-blog/components/header.js)
```
I think we want the absolute URL here?

LekoArts commented on examples/using-multiple-themes/src/gatsby-plugin-theme-ui/index.js:
```diff
@@ -0,0 +1,17 @@
+import merge from "deepmerge"
```
Maybe add a comment that `deepmerge` is available through gatsby-theme-blog?

laurieontech replied:
This is going to be updated shortly after a bug fix for theme-ui's "merge", so I think it's ok for now.

laurieontech commented:
Turns out this actually surfaced a bug with the merge functionality in theme-ui. So I'm going to move forward with this PR and once that's fixed we can go ahead and make the changes after the fact. Thanks again for all your hard work @ehowey!

ehowey commented:
Can you double check that the link to the example repo is working - I tried on my phone and got a 404.

laurieontech commented:
Ya, that one was my bad. I merged a fix to it right after publishing that should be live!
